Project: Enumerables finished.

MrSayyod opened the pull request:
- created a dedicated repository for this project;
- set up a code linter and stickler;
- followed the GitFlow guidelines;
- fixed linter errors;
- created descriptive README file;
- worked solo.

MrSayyod commented:
Added cops to rubocop.yml according to stickler recommendation:
Here is the stickler recommendation:
```
The following cops were added to RuboCop, but are not configured. Please set Enabled to either `true` or `false` in your `.rubocop.yml` file:
 - Lint/RaiseException (0.81)
 - Lint/StructNewOverride (0.81)
 - Style/HashEachMethods (0.80)
 - Style/HashTransformKeys (0.80)
 - Style/HashTransformValues (0.80)
```

pelzolga123 commented:
## Hi Sayyod! ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
### Your did an amazing job :+1: 
Just a few changes required before this project can be approved:
* [x] Don't forget to add a brief instruction on how to run your project in README file. 
* #### My_each
* [x] Should return an Enumerator if no block is given. For example:
`[1, 2, 3].my_each` - should return` #<Enumerator: [1, 2, 3]:my_each>`

*  #### My_each_with_index
* [x] Should return an Enumerator if no block is given. For example:
`[1, 2, 3].my_each_with_index` - should return` #<Enumerator: [1, 2, 3]:my_each_with_index>`

*  #### My_select
* [x] Should return an Enumerator if no block is given. For example: 
`[1, 2 ,3].my_select` - should return `#<Enumerator: [1, 2, 3]:my_select>`

*  #### My_map
* [x] Should return an Enumerator if no block is given. For example: 
`[1, 2 ,3].my_map` - should return `#<Enumerator: [1, 2, 3]:my_map>`

*  #### My_inject
* [x] When a symbol is specified should combine each element of the collection by applying the symbol as a named method. For example: 
`[1, 2, 3, 4, 5].my_inject(:+)` - should return 15

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

MrSayyod commented:
- on the Prerequisites part of the README.md file, I gave brief instructions to run the project;
- added "return to_enum" method to My_each, My_each_with_index, My_select and My_map methods;
- changed my_inject method according to TSE review

euqueme commented on enumerables.rb:
```diff
@@ -0,0 +1,96 @@
+    new_array
+  end
+
+  def my_inject(*par)
+    memo = par[0] if par[0].is_a? Integer
+    if par[0].is_a?(Symbol) || par[0].is_a?(String)
+      my_each { |element| memo = memo ? memo.send(par[0], element) : element }
+      memo
+    else
+      my_each { |element| memo = memo ? yield(memo, element) : element }
+    end
+    memo
+  end
```
- [x] should have working `my_inject`  identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [x] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving a NameError
![image](https://user-images.githubusercontent.com/52209067/79671469-a5c5e780-8187-11ea-9d06-f291c53d4f44.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/77698409-2ffaa180-6f76-11ea-91d3-9778aa560411.png)

- [x] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(2, :+) == [2, 5, 6].inject(2, :+) => 15
```
- Yours is retrieving a LocalJumpError
![image](https://user-images.githubusercontent.com/52209067/79671502-e0c81b00-8187-11ea-8afc-64d07788e63f.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/79671514-f0dffa80-8187-11ea-8a4f-27fae3f6e849.png)

MrSayyod replied:
Fixed

MrSayyod replied:
my_inject method working now

euqueme commented on enumerables.rb:
```diff
@@ -0,0 +1,96 @@
+    memo
+  end
+
+  def multiply_els
+    my_inject { |memo, element| memo * element }
+  end
```
[OPTIONAL]

- [x] mutiple_els is not an enumerable method so I strongly recommend putting it in a different module

MrSayyod replied:
Put it out from Enumerables module

euqueme commented on enumerables.rb:
```diff
@@ -0,0 +1,96 @@
+    condition
+  end
+
+  def my_none?(*par)
+    condition = true
+    if !par[0].nil?
+      my_each { |element| condition = false if par[0] === element } # rubocop:disable Style/CaseEquality
+    elsif !block_given?
+      my_each { |element| condition = false if element }
+    else
+      my_each { |element| condition = false if yield(element) }
+    end
+    condition
+  end
```
[OPTIONAL]

I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:
```ruby
  def my_none?(arg = nil, &block)
    !my_any?(arg, &block)
  end
```
To make this work we have to use block.call(element) instead of yield(element) in my_any?

MrSayyod replied:
Fixed! Thank you for help @euqeme!
Now my_none method is less redundant 👍 !

MrSayyod commented:
All changes are done!

majovanilla commented:
Hey @MrSayyod , 👋

## You did an *amazing*🌟 ✨ job on this project! 👊🏻  💯

All your methods are working as expected and your project has been **APPROVED**
You can merge your branch and continue with the next project!!

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes
